File: RxDemo/src/main/java/com/tamsiree/rxdemo/activity/ActivityContact.java
package com.tamsiree.rxdemo.activity;

import android.os.Bundle;

import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import com.tamsiree.rxdemo.R;
import com.tamsiree.rxtool.RxDeviceTool;
import com.tamsiree.rxui.activity.ActivityBase;
import com.tamsiree.rxui.model.ModelContactCity;
import com.tamsiree.rxui.view.RxTitle;
import com.tamsiree.rxui.view.wavesidebar.ComparatorLetter;
import com.tamsiree.rxui.view.wavesidebar.PinnedHeaderDecoration;
import com.tamsiree.rxui.view.wavesidebar.WaveSideBarView;
import com.tamsiree.rxui.view.wavesidebar.adapter.AdapterContactCity;

import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * @author tamsiree
 */
public class ActivityContact extends ActivityBase {

    @BindView(R.id.recycler_view)
    RecyclerView mRecyclerView;
    @BindView(R.id.side_view)
    WaveSideBarView mSideBarView;
    @BindView(R.id.rx_title)
    RxTitle mRxTitle;

    private AdapterContactCity mAdapterContactCity;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_contact);
        ButterKnife.bind(this);
        RxDeviceTool.setPortrait(this);
        initView();
    }

    private void initView() {
        mRxTitle.setLeftFinish(mContext);

        mRecyclerView.setLayoutManager(new LinearLayoutManager(this));

        final PinnedHeaderDecoration decoration = new PinnedHeaderDecoration();
        decoration.registerTypePinnedHeader(1, new PinnedHeaderDecoration.PinnedHeaderCreator() {
            @Override
            public boolean create(RecyclerView parent, int adapterPosition) {
                return true;
            }
        });
        mRecyclerView.addItemDecoration(decoration);


        new Thread(new Runnable() {
            @Override
            public void run() {
                Type listType = new TypeToken<ArrayList<ModelContactCity>>() {
                }.getType();
                Gson gson = new Gson();
                final List<ModelContactCity> list = gson.fromJson(ModelContactCity.DATA, listType);
                Collections.sort(list, new ComparatorLetter());
                runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        mAdapterContactCity = new AdapterContactCity(mContext, list);
                        mRecyclerView.setAdapter(mAdapterContactCity);
                    }
                });
            }
        }).start();

        mSideBarView.setOnTouchLetterChangeListener(new WaveSideBarView.OnTouchLetterChangeListener() {
            @Override
            public void onLetterChange(String letter) {
                int pos = mAdapterContactCity.getLetterPosition(letter);

                if (pos != -1) {
                    mRecyclerView.scrollToPosition(pos);
                    LinearLayoutManager mLayoutManager =
                            (LinearLayoutManager) mRecyclerView.getLayoutManager();
                    mLayoutManager.scrollToPositionWithOffset(pos, 0);
                }
            }
        });
    }
}

File: RxDemo/src/main/java/com/tamsiree/rxdemo/activity/ActivityContact.kt
package com.tamsiree.rxdemo.activity

import android.os.Bundle
import androidx.recyclerview.widget.LinearLayoutManager
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import com.tamsiree.rxdemo.R
import com.tamsiree.rxkit.RxDeviceTool.setPortrait
import com.tamsiree.rxui.activity.ActivityBase
import com.tamsiree.rxui.model.ModelContactCity
import com.tamsiree.rxui.view.wavesidebar.ComparatorLetter
import com.tamsiree.rxui.view.wavesidebar.PinnedHeaderDecoration
import com.tamsiree.rxui.view.wavesidebar.adapter.AdapterContactCity
import kotlinx.android.synthetic.main.activity_contact.*
import java.util.*

/**
 * @author tamsiree
 */
class ActivityContact : ActivityBase() {

    private var mAdapterContactCity: AdapterContactCity? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_contact)
        setPortrait(this)
    }

    override fun initView() {
        rx_title.setLeftFinish(mContext)
        recycler_view.layoutManager = LinearLayoutManager(this)
        val decoration = PinnedHeaderDecoration()
        decoration.registerTypePinnedHeader(1) { parent, adapterPosition -> true }
        recycler_view.addItemDecoration(decoration)

        side_view.setOnTouchLetterChangeListener { letter ->
            val pos = mAdapterContactCity!!.getLetterPosition(letter)
            if (pos != -1) {
                recycler_view.scrollToPosition(pos)
                val mLayoutManager = recycler_view.layoutManager as LinearLayoutManager?
                mLayoutManager!!.scrollToPositionWithOffset(pos, 0)
            }
        }
    }

    override fun initData() {
        Thread(Runnable {
            val listType = object : TypeToken<ArrayList<ModelContactCity?>?>() {}.type
            val gson = Gson()
            val list = gson.fromJson<List<ModelContactCity>>(ModelContactCity.DATA, listType)
            Collections.sort(list, ComparatorLetter())
            runOnUiThread {
                mAdapterContactCity = AdapterContactCity(mContext, list)
                recycler_view.adapter = mAdapterContactCity
            }
        }).start()
    }

}